klippy/extras/mixing_extruder.py

```diff
@@ -0,0 +1,335 @@
+                    for y in range(x + 1, steppers_count)][idx]
+    return [1. / steppers_count for p in range(steppers_count)]
+
+
+class MixingExtruder(object):
+    def __init__(self, config, parent=None):
+        self.printer = config.get_printer()
+        self._activated = False
+        self.name = "mixingextruder"
+        self.stepper_names = tuple(e.strip()
+                               for e in config.get('steppers', '').split(",")
+                               if len(e))
+        if not len(self.stepper_names):
+            raise self.printer.config_error(
+                "No steppers configured for mixing")
+        self._extruder_steppers = []
+        self._rotation_distances = ()
+        self._trapqs = ()
+        self.extruder_name = config.get('extruder_name',
+                                        self.stepper_names[0])
+        self._mixing = self._active_mixing = \
+            mixing_preset(0, len(self.stepper_names))
+        self._enabled = False
+
+        self.printer.register_event_handler("klippy:connect",
+                                            self._handle_connect)
+        # Register commands
+        gcode = self.printer.lookup_object('gcode')
+        gcode.register_command("SET_MIXING_EXTRUDER",
+                               self.cmd_SET_MIXING_EXTRUDER,
+                               desc=self.cmd_MIXING_STATUS_help)
+        gcode.register_command("MIXING_STATUS",
+                               self.cmd_MIXING_STATUS,
+                               desc=self.cmd_MIXING_STATUS_help)
+    def _handle_connect(self):
+        if self._activated:
+            return
+        self._activated = True
+        try:
+            for name in self.stepper_names:
+                try:
+                    stepper = (self.printer
+                               .lookup_object(name).extruder_stepper)
+                except:
+                    stepper = (self.printer
+                               .lookup_object("extruder_stepper " + name)
+                               .extruder_stepper)
+                self._extruder_steppers.append(stepper)
+        except Exception as e:
+            while len(self.steppers) > 0:
+                self.steppers.pop()
+            logging.error("no steppers found: %s: %s",
+                          ", ".join(self.stepper_names), e)
+        self._rotation_distances = tuple(
+            stepper.stepper.get_rotation_distance()[0]
+            for stepper in self._extruder_steppers)
+        self._trapqs = tuple(
+            stepper.stepper.get_trapq()
+            for stepper in self._extruder_steppers)
+    def get_status(self, eventtime):
+        return dict(
+            mixing_enabled="enabled" if self._enabled else "disabled",
+            mixing=",".join(
+                "%0.1f%%" % (m * 100.) for m in self._active_mixing),
+            mixing_ticks=",".join(
+                "%d" % (extruder_stepper.stepper.get_mcu_position())
```

Comment:
This looks to me like it is "peeking" into a member of the extruder_stepper object.  As per the Klipper conventions documented in docs/Code_Overview.md this is not valid.

Reply:
I have refactored the extruder_stepper/stepper access and as I understand get_mcu_position() is a "public" member of the stepper